File: rust/crates/console-log/src/cdp.rs
//! CDP WebSocket connection for console log streaming.
//!
//! Connects to Steam's CEF debugger and subscribes to `Runtime.consoleAPICalled`
//! and `Log.entryAdded` events. Parses `%c` CSS styled segments.

use std::time::Duration;

use capydeploy_protocol::console_log::{ConsoleLogEntry, StyledSegment};
use capydeploy_protocol::constants::log_level_bit;
use capydeploy_steam::cef::CefClient;
use futures_util::{SinkExt, StreamExt};
use serde::Deserialize;
use tokio_tungstenite::tungstenite::protocol::Message as WsMessage;

/// Timeout for the WebSocket handshake.
const WS_HANDSHAKE_TIMEOUT: Duration = Duration::from_secs(5);

/// A raw CDP event received from the browser.
#[derive(Debug, Deserialize)]
struct CdpEvent {
    method: Option<String>,
    params: Option<serde_json::Value>,
}

/// CDP message sent to the browser.
#[derive(serde::Serialize)]
struct CdpCommand {
    id: i32,
    method: &'static str,
    params: serde_json::Value,
}

/// `Runtime.consoleAPICalled` parameters.
#[derive(Deserialize)]
struct ConsoleApiCalledParams {
    #[serde(rename = "type")]
    call_type: String,
    args: serde_json::Value,
}

/// `Log.entryAdded` parameters.
#[derive(Deserialize)]
struct LogEntryAddedParams {
    entry: LogEntry,
}

#[derive(Deserialize)]
struct LogEntry {
    #[serde(default)]
    source: String,
    #[serde(default)]
    level: String,
    #[serde(default)]
    text: String,
    #[serde(default)]
    url: String,
    #[serde(default, rename = "lineNumber")]
    line: i32,
}

/// Discovers the CEF debug tab and returns its WebSocket URL.
pub async fn discover_ws_url() -> Result<String, String> {
    let client = CefClient::new();
    let tabs = client
        .get_tabs()
        .await
        .map_err(|e| format!("failed to discover CEF tabs: {e}"))?;

    let tab = CefClient::find_js_context(&tabs).map_err(|e| format!("no suitable tab: {e}"))?;

    if tab.web_socket_debugger_url.is_empty() {
        return Err("tab has no WebSocket debugger URL".into());
    }

    Ok(tab.web_socket_debugger_url.clone())
}

/// Result of connecting and streaming: why the connection ended.
pub enum StreamEnd {
    /// Context was cancelled (graceful shutdown).
    Cancelled,
    /// Connection or protocol error.
    Error(String),
}

/// Connects to a CEF tab's WebSocket, enables `Runtime` + `Log` domains,
/// and streams console events until the cancellation token fires or an error occurs.
///
/// Each parsed entry is passed to `on_entry` with its level bit already resolved.
pub async fn stream_events<F>(
    ws_url: &str,
    level_mask: impl Fn() -> u32,
    cancel: tokio_util::sync::CancellationToken,
    on_entry: F,
) -> StreamEnd
where
    F: Fn(ConsoleLogEntry),
{
    // Connect WebSocket.
    let ws = match tokio::time::timeout(
        WS_HANDSHAKE_TIMEOUT,
        tokio_tungstenite::connect_async(ws_url),
    )
    .await
    {
        Ok(Ok((ws, _))) => ws,
        Ok(Err(e)) => return StreamEnd::Error(format!("WebSocket connect failed: {e}")),
        Err(_) => return StreamEnd::Error("WebSocket handshake timeout".into()),
    };

    let (mut write, mut read) = ws.split();

    // Enable Runtime and Log domains.
    for (id, method) in [(1, "Runtime.enable"), (2, "Log.enable")] {
        let cmd = CdpCommand {
            id,
            method,
            params: serde_json::json!({}),
        };
        let json = match serde_json::to_string(&cmd) {
            Ok(j) => j,
            Err(e) => return StreamEnd::Error(format!("failed to serialize CDP command: {e}")),
        };
        if let Err(e) = write.send(WsMessage::Text(json.into())).await {
            return StreamEnd::Error(format!("failed to send CDP command: {e}"));
        }
    }

    tracing::debug!("CDP console log streaming started");

    // Read loop.
    loop {
        tokio::select! {
            _ = cancel.cancelled() => {
                let _ = write.close().await;
                return StreamEnd::Cancelled;
            }
            frame = read.next() => {
                match frame {
                    Some(Ok(WsMessage::Text(text))) => {
                        if let Ok(event) = serde_json::from_str::<CdpEvent>(&text)
                            && let Some(entry) = parse_event(&event, &level_mask)
                        {
                            on_entry(entry);
                        }
                    }
                    Some(Ok(WsMessage::Close(_))) | None => {
                        return StreamEnd::Error("CDP WebSocket closed".into());
                    }
                    Some(Err(e)) => {
                        return StreamEnd::Error(format!("CDP WebSocket read error: {e}"));
                    }
                    _ => {} // Ping/Pong/Binary — ignore.
                }
            }
        }
    }
}

/// Parses a CDP event into a `ConsoleLogEntry`, applying level filtering.
fn parse_event(event: &CdpEvent, level_mask: &impl Fn() -> u32) -> Option<ConsoleLogEntry> {
    let method = event.method.as_deref()?;
    let params = event.params.as_ref()?;

    match method {
        "Runtime.consoleAPICalled" => {
            let api: ConsoleApiCalledParams = serde_json::from_value(params.clone()).ok()?;
            let level = normalize_level(&api.call_type);
            let bit = log_level_bit(&level);
            if bit == 0 || level_mask() & bit == 0 {
                return None;
            }
            let (text, segments) = format_console_args(&api.args);
            Some(ConsoleLogEntry {
                timestamp: now_millis(),
                level,
                source: "console".into(),
                text,
                url: String::new(),
                line: 0,
                segments,
            })
        }
        "Log.entryAdded" => {
            let log: LogEntryAddedParams = serde_json::from_value(params.clone()).ok()?;
            let entry = &log.entry;
            let level = normalize_level(&entry.level);
            let bit = log_level_bit(&level);
            if bit == 0 || level_mask() & bit == 0 {
                return None;
            }
            Some(ConsoleLogEntry {
                timestamp: now_millis(),
                level,
                source: entry.source.clone(),
                text: entry.text.clone(),
                url: entry.url.clone(),
                line: entry.line,
                segments: vec![],
            })
        }
        _ => None,
    }
}

/// Normalizes CDP level strings to our canonical names.
fn normalize_level(level: &str) -> String {
    match level {
        "warning" => "warn".into(),
        "verbose" => "debug".into(),
        other => other.to_string(),
    }
}

/// Formats console API arguments, handling `%c` styled segments.
///
/// If the first argument contains `%c` directives, subsequent arguments are
/// consumed as CSS styles for the corresponding text segments.
fn format_console_args(args: &serde_json::Value) -> (String, Vec<StyledSegment>) {
    let arr = match args.as_array() {
        Some(a) if !a.is_empty() => a,
        _ => return (String::new(), vec![]),
    };

    // Extract text value from a console argument object.
    let arg_text = |arg: &serde_json::Value| -> String {
        if let Some(v) = arg.get("value") {
            match v {
                serde_json::Value::String(s) => s.clone(),
                serde_json::Value::Null => "null".into(),
                other => other.to_string(),
            }
        } else if let Some(desc) = arg.get("description").and_then(|d| d.as_str()) {
            desc.to_string()
        } else {
            String::new()
        }
    };

    let first_text = arg_text(&arr[0]);

    // Check if first arg contains %c directives.
    if !first_text.contains("%c") {
        // Plain text: concatenate all args.
        let text: String = arr.iter().map(arg_text).collect::<Vec<_>>().join(" ");
        return (text, vec![]);
    }

    // Split by %c and build styled segments.
    let parts: Vec<&str> = first_text.split("%c").collect();
    let mut segments = Vec::new();
    let mut css_arg_idx = 1usize; // CSS args start at index 1.

    for (i, part) in parts.iter().enumerate() {
        if i == 0 {
            // Text before the first %c — no CSS.
            if !part.is_empty() {
                segments.push(StyledSegment {
                    text: part.to_string(),
                    css: String::new(),
                });
            }
        } else {
            // Text after %c[i], consumes args[css_arg_idx] as CSS.
            let css = if css_arg_idx < arr.len() {
                let css_val = arg_text(&arr[css_arg_idx]);
                css_arg_idx += 1;
                css_val
            } else {
                String::new()
            };
            if !part.is_empty() || !css.is_empty() {
                segments.push(StyledSegment {
                    text: part.to_string(),
                    css,
                });
            }
        }
    }

    // Remaining args after all CSS values are appended as plain text.
    for arg in arr.iter().skip(css_arg_idx) {
        let text = arg_text(arg);
        if !text.is_empty() {
            segments.push(StyledSegment {
                text,
                css: String::new(),
            });
        }
    }

    // Build plain text from all segments.
    let plain: String = segments.iter().map(|s| s.text.as_str()).collect();

    (plain, segments)
}

/// Current timestamp in milliseconds.
fn now_millis() -> i64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_millis() as i64)
        .unwrap_or(0)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn normalize_level_mapping() {
        assert_eq!(normalize_level("warning"), "warn");
        assert_eq!(normalize_level("verbose"), "debug");
        assert_eq!(normalize_level("log"), "log");
        assert_eq!(normalize_level("error"), "error");
        assert_eq!(normalize_level("info"), "info");
    }

    #[test]
    fn format_plain_args() {
        let args = serde_json::json!([
            {"type": "string", "value": "hello"},
            {"type": "string", "value": "world"}
        ]);
        let (text, segments) = format_console_args(&args);
        assert_eq!(text, "hello world");
        assert!(segments.is_empty());
    }

    #[test]
    fn format_styled_args() {
        // console.log("%cHello%c World", "color: red", "color: blue")
        let args = serde_json::json!([
            {"type": "string", "value": "%cHello%c World"},
            {"type": "string", "value": "color: red"},
            {"type": "string", "value": "color: blue"}
        ]);
        let (text, segments) = format_console_args(&args);
        assert_eq!(text, "Hello World");
        assert_eq!(segments.len(), 2);
        assert_eq!(segments[0].text, "Hello");
        assert_eq!(segments[0].css, "color: red");
        assert_eq!(segments[1].text, " World");
        assert_eq!(segments[1].css, "color: blue");
    }

    #[test]
    fn format_styled_with_trailing_args() {
        // console.log("%cStyled", "color: red", "extra text")
        let args = serde_json::json!([
            {"type": "string", "value": "%cStyled"},
            {"type": "string", "value": "color: red"},
            {"type": "string", "value": "extra text"}
        ]);
        let (text, segments) = format_console_args(&args);
        assert_eq!(text, "Styledextra text");
        assert_eq!(segments.len(), 2);
        assert_eq!(segments[0].css, "color: red");
        assert_eq!(segments[1].text, "extra text");
        assert!(segments[1].css.is_empty());
    }

    #[test]
    fn format_empty_args() {
        let args = serde_json::json!([]);
        let (text, segments) = format_console_args(&args);
        assert!(text.is_empty());
        assert!(segments.is_empty());
    }

    #[test]
    fn format_non_string_args() {
        let args = serde_json::json!([
            {"type": "number", "value": 42},
            {"type": "boolean", "value": true},
            {"type": "object", "description": "[object Object]"}
        ]);
        let (text, segments) = format_console_args(&args);
        assert_eq!(text, "42 true [object Object]");
        assert!(segments.is_empty());
    }

    #[test]
    fn format_null_value() {
        let args = serde_json::json!([
            {"type": "object", "subtype": "null", "value": null}
        ]);
        let (text, _) = format_console_args(&args);
        assert_eq!(text, "null");
    }

    #[test]
    fn parse_console_api_event() {
        let event = CdpEvent {
            method: Some("Runtime.consoleAPICalled".into()),
            params: Some(serde_json::json!({
                "type": "log",
                "args": [{"type": "string", "value": "test message"}]
            })),
        };
        let mask = || 0xFFu32;
        let entry = parse_event(&event, &mask).unwrap();
        assert_eq!(entry.level, "log");
        assert_eq!(entry.source, "console");
        assert_eq!(entry.text, "test message");
        assert!(entry.timestamp > 0);
    }

    #[test]
    fn parse_log_entry_added_event() {
        let event = CdpEvent {
            method: Some("Log.entryAdded".into()),
            params: Some(serde_json::json!({
                "entry": {
                    "source": "javascript",
                    "level": "error",
                    "text": "Uncaught TypeError",
                    "url": "https://example.com/app.js",
                    "lineNumber": 42
                }
            })),
        };
        let mask = || 0xFFu32;
        let entry = parse_event(&event, &mask).unwrap();
        assert_eq!(entry.level, "error");
        assert_eq!(entry.source, "javascript");
        assert_eq!(entry.text, "Uncaught TypeError");
        assert_eq!(entry.url, "https://example.com/app.js");
        assert_eq!(entry.line, 42);
    }

    #[test]
    fn parse_event_filters_by_mask() {
        let event = CdpEvent {
            method: Some("Runtime.consoleAPICalled".into()),
            params: Some(serde_json::json!({
                "type": "debug",
                "args": [{"type": "string", "value": "debug msg"}]
            })),
        };
        // Default mask (15) excludes debug (16).
        let mask = || capydeploy_protocol::constants::LOG_LEVEL_DEFAULT;
        let entry = parse_event(&event, &mask);
        assert!(entry.is_none(), "debug should be filtered out");
    }

    #[test]
    fn parse_event_passes_debug_when_enabled() {
        let event = CdpEvent {
            method: Some("Runtime.consoleAPICalled".into()),
            params: Some(serde_json::json!({
                "type": "debug",
                "args": [{"type": "string", "value": "debug msg"}]
            })),
        };
        let mask = || 0xFFu32; // All levels enabled.
        let entry = parse_event(&event, &mask);
        assert!(entry.is_some());
    }

    #[test]
    fn parse_unknown_event_returns_none() {
        let event = CdpEvent {
            method: Some("Page.loadEventFired".into()),
            params: Some(serde_json::json!({})),
        };
        let mask = || 0xFFu32;
        assert!(parse_event(&event, &mask).is_none());
    }

    #[test]
    fn parse_event_normalizes_warning() {
        let event = CdpEvent {
            method: Some("Log.entryAdded".into()),
            params: Some(serde_json::json!({
                "entry": {
                    "source": "network",
                    "level": "warning",
                    "text": "resource not found"
                }
            })),
        };
        let mask = || 0xFFu32;
        let entry = parse_event(&event, &mask).unwrap();
        assert_eq!(entry.level, "warn");
    }
}

File: deprecated/apps/agents/desktop/consolelog/cdp.go
// Package consolelog streams console logs from Steam's CEF debugger via CDP.
package consolelog

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"sync"
	"time"

	"github.com/gorilla/websocket"
	"github.com/lobinuxsoft/capydeploy/pkg/protocol"
)

const (
	cefDebugEndpoint = "http://localhost:8080/json"
	cefHTTPTimeout   = 5 * time.Second
	cefWSHandshake   = 5 * time.Second
)

// cdpConn is a persistent WebSocket connection to a CDP endpoint.
type cdpConn struct {
	conn     *websocket.Conn
	msgID    int
	closeOnce sync.Once
}

type cefTab struct {
	Title                string `json:"title"`
	Type                 string `json:"type"`
	WebSocketDebuggerURL string `json:"webSocketDebuggerUrl"`
}

// cdpEvent represents a raw CDP event or response.
type cdpEvent struct {
	ID     int             `json:"id,omitempty"`
	Method string          `json:"method,omitempty"`
	Params json.RawMessage `json:"params,omitempty"`
}

// consoleAPICalledParams matches Runtime.consoleAPICalled event.
type consoleAPICalledParams struct {
	Type string          `json:"type"`
	Args json.RawMessage `json:"args"`
}

// consoleArg represents a single argument in Runtime.consoleAPICalled.
type consoleArg struct {
	Type  string `json:"type"`
	Value any    `json:"value,omitempty"`
	// For string values, description may also be present
	Description string `json:"description,omitempty"`
}

// logEntryAddedParams matches Log.entryAdded event.
type logEntryAddedParams struct {
	Entry logEntry `json:"entry"`
}

type logEntry struct {
	Source string `json:"source"`
	Level  string `json:"level"`
	Text   string `json:"text"`
	URL    string `json:"url,omitempty"`
	Line   int    `json:"lineNumber,omitempty"`
}

// getTabs fetches the list of debuggable tabs from CEF.
func getTabs(ctx context.Context) ([]cefTab, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, cefDebugEndpoint, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	client := &http.Client{Timeout: cefHTTPTimeout}
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to connect to CEF debugger: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read CEF response: %w", err)
	}

	var tabs []cefTab
	if err := json.Unmarshal(body, &tabs); err != nil {
		return nil, fmt.Errorf("failed to parse CEF tabs: %w", err)
	}

	return tabs, nil
}

// findJSContext finds the SharedJSContext or SP tab for console log subscription.
func findJSContext(tabs []cefTab) (*cefTab, error) {
	var spTab *cefTab

	for i := range tabs {
		tab := &tabs[i]
		if tab.WebSocketDebuggerURL == "" {
			continue
		}
		if tab.Title == "SharedJSContext" {
			return tab, nil
		}
		if tab.Title == "SP" && spTab == nil {
			spTab = tab
		}
	}

	if spTab != nil {
		return spTab, nil
	}

	return nil, fmt.Errorf("no suitable JS context found (need SharedJSContext or SP tab)")
}

// dialCDP connects to a CDP WebSocket endpoint.
func dialCDP(ctx context.Context) (*cdpConn, error) {
	tabs, err := getTabs(ctx)
	if err != nil {
		return nil, err
	}

	tab, err := findJSContext(tabs)
	if err != nil {
		return nil, err
	}

	dialer := websocket.Dialer{
		HandshakeTimeout: cefWSHandshake,
	}

	conn, _, err := dialer.DialContext(ctx, tab.WebSocketDebuggerURL, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to connect to CDP: %w", err)
	}

	return &cdpConn{conn: conn}, nil
}

// sendCommand sends a CDP command and returns its ID.
func (c *cdpConn) sendCommand(method string, params map[string]any) (int, error) {
	c.msgID++
	msg := map[string]any{
		"id":     c.msgID,
		"method": method,
	}
	if params != nil {
		msg["params"] = params
	}

	if err := c.conn.WriteJSON(msg); err != nil {
		return 0, fmt.Errorf("failed to send CDP command %s: %w", method, err)
	}
	return c.msgID, nil
}

// enableRuntime sends Runtime.enable to receive consoleAPICalled events.
func (c *cdpConn) enableRuntime() error {
	_, err := c.sendCommand("Runtime.enable", nil)
	return err
}

// enableLog sends Log.enable to receive entryAdded events.
func (c *cdpConn) enableLog() error {
	_, err := c.sendCommand("Log.enable", nil)
	return err
}

// readEvent reads the next CDP event from the WebSocket.
// Returns the raw event for the caller to dispatch.
func (c *cdpConn) readEvent() (*cdpEvent, error) {
	var event cdpEvent
	if err := c.conn.ReadJSON(&event); err != nil {
		return nil, err
	}
	return &event, nil
}

// close closes the CDP WebSocket connection. Safe to call multiple times.
func (c *cdpConn) close() {
	c.closeOnce.Do(func() {
		c.conn.Close()
	})
}

// consoleArgsResult holds parsed console args with optional styled segments.
type consoleArgsResult struct {
	Text     string
	Segments []protocol.StyledSegment
}

// formatConsoleArgsRich converts Runtime.consoleAPICalled args to text + styled segments.
// Detects %c format directives and splits into segments with CSS.
func formatConsoleArgsRich(raw json.RawMessage) consoleArgsResult {
	var args []consoleArg
	if err := json.Unmarshal(raw, &args); err != nil {
		return consoleArgsResult{Text: string(raw)}
	}

	if len(args) == 0 {
		return consoleArgsResult{}
	}

	// Check if first arg is a string containing %c
	firstStr := ""
	if args[0].Type == "string" {
		if s, ok := args[0].Value.(string); ok {
			firstStr = s
		}
	}

	if firstStr == "" || !strings.Contains(firstStr, "%c") {
		// No %c formatting — produce plain text
		text := formatConsoleArgsPlain(args)
		return consoleArgsResult{Text: text}
	}

	// Split by %c, consume CSS args in order
	parts := strings.Split(firstStr, "%c")
	var segments []protocol.StyledSegment
	cssArgIdx := 1 // CSS args start at index 1

	for i, part := range parts {
		if i == 0 && part == "" {
			continue // leading empty before first %c
		}
		if i == 0 {
			// Text before the first %c — no CSS
			segments = append(segments, protocol.StyledSegment{Text: part})
		} else {
			css := ""
			if cssArgIdx < len(args) {
				css = formatArg(args[cssArgIdx])
			}
			cssArgIdx++
			segments = append(segments, protocol.StyledSegment{Text: part, CSS: css})
		}
	}

	// Append remaining non-CSS args as plain text
	for i := cssArgIdx; i < len(args); i++ {
		s := formatArg(args[i])
		if s != "" {
			segments = append(segments, protocol.StyledSegment{Text: " " + s})
		}
	}

	// Build plain text from all segments
	var sb strings.Builder
	for _, seg := range segments {
		sb.WriteString(seg.Text)
	}

	return consoleArgsResult{Text: sb.String(), Segments: segments}
}

// formatConsoleArgsPlain converts args to a plain string (no %c handling).
func formatConsoleArgsPlain(args []consoleArg) string {
	if len(args) == 1 {
		return formatArg(args[0])
	}
	result := ""
	for i, arg := range args {
		if i > 0 {
			result += " "
		}
		result += formatArg(arg)
	}
	return result
}

func formatArg(arg consoleArg) string {
	if arg.Description != "" {
		return arg.Description
	}
	if arg.Type == "string" {
		if s, ok := arg.Value.(string); ok {
			return s
		}
	}
	if arg.Value != nil {
		b, _ := json.Marshal(arg.Value)
		return string(b)
	}
	return fmt.Sprintf("[%s]", arg.Type)
}
